crypto/sm2/sm2_crypt.c

```diff
@@ -0,0 +1,262 @@
+        ASN1_SIMPLE(SM2_Ciphertext, C2,
+                        ASN1_OCTET_STRING),} ASN1_SEQUENCE_END(SM2_Ciphertext)
+IMPLEMENT_ASN1_FUNCTIONS(SM2_Ciphertext)
+
+static int EC_field_size(const EC_GROUP *group)
+{
+    /* Is there some simpler way to do this? */
+    BIGNUM *p = BN_new();
+    BIGNUM *a = BN_new();
+    BIGNUM *b = BN_new();
+    int field_size;
+
+    EC_GROUP_get_curve_GFp(group, p, a, b, NULL);
+    field_size = (BN_num_bits(p) + 7) / 8;
+
+    BN_free(p);
+    BN_free(a);
+    BN_free(b);
+
+    return field_size;
+}
+
+int SM2_ciphertext_size(const EC_KEY *key, const EVP_MD *digest, size_t msg_len)
+{
+    return 10 + 2 * EC_field_size(EC_KEY_get0_group(key)) +
+        EVP_MD_size(digest) + msg_len;
+}
+
+int SM2_encrypt(const EC_KEY *key,
+                const EVP_MD *digest,
+                const uint8_t *msg,
+                size_t msg_len, uint8_t *ciphertext_buf, size_t *ciphertext_len)
+{
+    int rc = 0;
+    BIGNUM *k = BN_new();
+    BIGNUM *x1 = BN_new();
+    BIGNUM *y1 = BN_new();
+    BIGNUM *x2 = BN_new();
+    BIGNUM *y2 = BN_new();
+
+    BN_CTX *ctx = BN_CTX_new();
+    EVP_MD_CTX *hash = EVP_MD_CTX_new();
+
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+    const BIGNUM *order = EC_GROUP_get0_order(group);
+    EC_POINT *kG = EC_POINT_new(group);
+    EC_POINT *kP = EC_POINT_new(group);
+    const EC_POINT *P = EC_KEY_get0_public_key(key);
+    uint8_t msg_mask[msg_len];
+
+    const size_t field_size = EC_field_size(group);
+    uint8_t x2y2[2 * field_size];
+    uint8_t x2_bits[field_size];
+    uint8_t y2_bits[field_size];
+
+    const size_t C3_size = EVP_MD_size(digest);
+    uint8_t C3[C3_size];
+
+    memset(ciphertext_buf, 0, *ciphertext_len);
+
+    BN_priv_rand_range(k, order);
+
+    if (EC_POINT_mul(group, kG, k, NULL, NULL, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_get_affine_coordinates_GFp(group, kG, x1, y1, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_mul(group, kP, NULL, P, k, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_get_affine_coordinates_GFp(group, kP, x2, y2, ctx) == 0)
+        goto done;
+
+    BN_bn2binpad(x2, x2_bits, field_size);
+    BN_bn2binpad(y2, y2_bits, field_size);
+
+    memcpy(x2y2, x2_bits, field_size);
+    memcpy(x2y2 + field_size, y2_bits, field_size);
+
+    // This happens to match the KDF used in SM2
+    if (ECDH_KDF_X9_62(msg_mask, msg_len, x2y2, 2 * field_size, NULL, 0, digest)
+        == 0)
+        goto done;
+
+    for (size_t i = 0; i != msg_len; ++i)
+        msg_mask[i] ^= msg[i];
+
+    if (EVP_DigestInit(hash, digest) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, x2_bits, field_size) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, msg, msg_len) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, y2_bits, field_size) == 0)
+        goto done;
+
+    if (EVP_DigestFinal(hash, C3, NULL) == 0)
+        goto done;
+
+    struct SM2_Ciphertext_st ctext_struct;
```

Comment:
We require ANSI C syntax, so this definition must move up to the top of the function.